kpatch-build/create-diff-object.c

```diff
@@ -509,6 +509,74 @@ static int kpatch_line_macro_change_only(struct section *sec)
 
 	return 1;
 }
+#elif __powerpc__
+#define PPC_INSTR_LEN 4
+#define PPC_RA_OFFSET 16
+
+static int kpatch_line_macro_change_only(struct section *sec)
+{
+	unsigned long start1, start2, size, offset;
+	unsigned int instr1, instr2;
+	struct rela *rela;
+	int lineonly = 0, found;
+
+	if (sec->status != CHANGED ||
+	    is_rela_section(sec) ||
+	    !is_text_section(sec) ||
+	    sec->sh.sh_size != sec->twin->sh.sh_size ||
+	    !sec->rela ||
+	    sec->rela->status != SAME)
+		return 0;
+
+	start1 = (unsigned long)sec->twin->data->d_buf;
+	start2 = (unsigned long)sec->data->d_buf;
+	size = sec->sh.sh_size;
+	for (offset = 0; offset < size; offset += PPC_INSTR_LEN) {
+		if (!memcmp((void *)start1 + offset, (void *)start2 + offset,
+			    PPC_INSTR_LEN))
+			continue;
+
+		instr1 = *(unsigned int *)(start1 + offset) >> PPC_RA_OFFSET;
+		instr2 = *(unsigned int *)(start2 + offset) >> PPC_RA_OFFSET;
+
+		/* verify it's a load immediate to r5 */
+		if (!(instr1 == 0x38a0 && instr2 == 0x38a0))
+			return 0;
+
+		found = 0;
+		list_for_each_entry(rela, &sec->rela->relas, list) {
+			if (rela->offset < offset + PPC_INSTR_LEN)
+				continue;
+			if (rela->string)
+				continue;
+			if (!strncmp(rela->sym->name, "__warned.", 9))
+				continue;
+			if (!strcmp(rela->sym->name, ".toc"))
+				continue;
```

Comment:
Instead of checking for a .toc rela, I think it would be good to be more specific here.  When you saw this, what was the "second level" rela, i.e. what did toc_rela(rela) refer to?

Reply:
For all relas toc_rela was refering to a rela with sym->name ".rodata.netdev_is_rx_handler_busy.str1.8". Will "second level" rela have toc_rela() as non-null and we can just check for that?

Reply:
Ah, that's a string, so I think you can just change the `if (rela->string)` to `if (toc_rela(rela)->string)`.